### PR TITLE
Addressbook migration

### PR DESCRIPTION
**Pull request by Nana-EC**
**Detailed description**:
Streamline address book migration logic to increase code sharing, testing but also allow for migration of file_data -> address_book to occur on demand when needed.

- Add a `migrate()` method to the AddressBookService interface
- Move migration logic from `V1_28_1__Address_Book` to `AddressBookServiceImpl`
- Remove `V1_28_1__Address_Book` in favor of on demand `AddressBookServiceImpl.migrate()`
- Add a `MirrorBaseJavaMigration` abstract class to hold shared java migration logic. In this case add a `skipMigrationVersion()` that ensures migrations prior to baseline are not run. This is in anticipation of v2 db schemas where v1 should not be run
- Update `V1_11_6__Missing_Entities`  to extend `MirrorBaseJavaMigration`
- Update `FileDataRepository` w `findAddressBooksInRange()` which was original sql migration query
- Update `AddressBookServiceImpl.getCurrent()` to handle empty table case and call inner migration logic to bootstrap initial filesystem/classpath address book as well as bring in historic addressBooks
- Update `AddressBookServiceImpl.update()` to ensure address_book` table is populated prior to adding any new address books
- Update tests that allowed for no address book after calling `getCurrent()` or `update()` no initial address book given each test clears the results of migrations

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:
This is fallout from comments on PR #1319 and also some desired improvements.
I would like to get this in first to reduce foot print of above PR

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 1]
```diff
@@ -111,6 +136,17 @@ public boolean isAddressBook(EntityId entityId) {
         return ADDRESS_BOOK_101_ENTITY_ID.equals(entityId) || ADDRESS_BOOK_102_ENTITY_ID.equals(entityId);
     }
 
+    @Override
+    public void migrate() {
+        log.debug("Searching for address book on file system");
+        AddressBook addressBook = buildAddressBook(getInitialAddressBookFileData());
+        addressBook = addressBookRepository.save(addressBook);
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:143]: `migrate()` needs to be in a transaction to be functionally equivalent to the flyway approach. What's less clear if it's okay to put `@Transactional` on the public methods of this class and let it join outer transactions or it requires its own inner transaction via `TransactionTemplate`. I'm thinking the latter.

Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:143]: Technically `buildAddressBook()` has the potential to return null. Why not reuse `parse()` method?

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 2]
```diff
@@ -269,4 +303,92 @@ private void updatePreviousAddressBook(FileData fileData) {
...
+        log.info("Processed {} historic address books", fileDataEntries.get());
+    }
+
+    /**
+     * Run address book migration logic and return the latest addressBook 102 object
+     *
+     * @param consensusTimestamp time stamp reference to get latest address book
+     * @return
+     */
+    private AddressBook migrateAndRetrieveLatestAddressBook(long consensusTimestamp) {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:387]: This method seems unnecessary. You've already retrieved all the data at some point inside the `migrate()` method. Just have it keep track of which is the latest and return that so the signature becomes:
```java
private synchronized AddressBook migrate() {
...
}
```

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_28_1__Address_Book.java [hunk 3]
```diff
@@ -21,115 +21,32 @@
...
-                    @Override
-                    public FileData mapRow(ResultSet rs, int rowNum) throws SQLException {
-                        return new FileData(rs.getLong("consensus_timestamp"),
-                                rs.getBytes("file_data"),
-                                EntityIdEndec.decode(rs.getInt("entity_id"), EntityTypeEnum.FILE),
-                                rs.getInt("transaction_type"));
-                    }
-                });
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        addressBookService.migrate();
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_28_1__Address_Book.java:48]: Good point
Given Java migration checksums are null and we have `ignoreMissingMigrations` set to true should be fine to remove this.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_11_6__Missing_Entities.java [hunk 4]
```diff
@@ -54,14 +53,18 @@
...
-            @Lazy JdbcOperations jdbcOperations) {
+                                     @Lazy JdbcOperations jdbcOperations) {
         this.mirrorProperties = mirrorProperties;
         this.entityRepository = entityRepository;
         this.jdbcOperations = jdbcOperations;
     }
 
     @Override
     public void migrate(Context context) throws Exception {
+        if (skipMigrationVersion(getVersion(), context.getConfiguration())) {
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_11_6__Missing_Entities.java:64]: Good suggestion. Will make this change

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:387]: Had considered that, went with the above to be more explicit but I'm okay with this approach.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 5]
```diff
@@ -82,6 +97,11 @@ public void update(FileData fileData) {
             return;
         }
 
+        // ensure address_book table is populated with at least bootstrap addressBook prior to additions
+        if (shouldMigrate()) {
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:101]: Works for me

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 6]
```diff
@@ -111,6 +136,17 @@ public boolean isAddressBook(EntityId entityId) {
         return ADDRESS_BOOK_101_ENTITY_ID.equals(entityId) || ADDRESS_BOOK_102_ENTITY_ID.equals(entityId);
     }
 
+    @Override
+    public void migrate() {
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:140]: Good point. I'll move the synchronized back up here instead of only on the file read method

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:143]: Will go with `TransactionTemplate` to be safe

Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:143]: Yeah `parse` is better. Less code as it does what I need.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 7]
```diff
@@ -269,4 +303,92 @@ private void updatePreviousAddressBook(FileData fileData) {
...
+
+    /**
+     * Batch parse all 101 and 102 addressBook fileData objects and update the address_book table
+     */
+    private void parseHistoricAddressBooks() {
+        AtomicLong fileDataEntries = new AtomicLong(0);
+        AtomicLong currentConsensusTimestamp = new AtomicLong(0);
+
+        // starting from consensusTimeStamp = 0 retrieve pages of fileData entries for historic address books
+        int pageSize = 1000; //
```
Comment (ijungmann) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:360]: Nit, unless there's supposed to be a comment here:
```suggestion
        int pageSize = 1000;
```

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_11_6__Missing_Entities.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_11_6__Missing_Entities.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_28_1__Address_Book.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_28_1__Address_Book.java]: Removed

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Modified to call `migrate()` directly and removed `shouldMigrate()`

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Done

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Added `TransactionTemplate` execute

Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+import javax.inject.Named;
+import lombok.extern.log4j.Log4j2;
+import org.flywaydb.core.api.MigrationVersion;
+import org.flywaydb.core.api.configuration.Configuration;
+import org.flywaydb.core.api.migration.BaseJavaMigration;
+
+@Log4j2
+@Named
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java]: Removed

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 7 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:367]: Will remove thanks

Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 9]
```diff
@@ -95,9 +116,14 @@ public void update(FileData fileData) {
...
         Instant now = Instant.now();
         long consensusTimestamp = Utility.convertToNanosMax(now.getEpochSecond(), now.getNano());
+
+        // retrieve latest address book. If address_book is empty parse initial and historic address book files
         return addressBookRepository
                 .findLatestAddressBook(consensusTimestamp, ADDRESS_BOOK_102_ENTITY_ID.getId())
-                .orElseThrow(() -> new IllegalStateException("No valid address book found in DB"));
+                .orElseGet(() -> {
+                    log.debug("No address book found in db");
+                    return migrate();
```
Comment (ijungmann) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:125]: Iis there any risk here of `migrate()` returning null?  I just see that `findLatestAddressBook()` is a more granular search than the one `migrate()` does.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 10]
```diff
@@ -95,9 +116,14 @@ public void update(FileData fileData) {
     public AddressBook getCurrent() {
         Instant now = Instant.now();
         long consensusTimestamp = Utility.convertToNanosMax(now.getEpochSecond(), now.getNano());
+
+        // retrieve latest address book. If address_book is empty parse initial and historic address book files
         return addressBookRepository
                 .findLatestAddressBook(consensusTimestamp, ADDRESS_BOOK_102_ENTITY_ID.getId())
-                .orElseThrow(() -> new IllegalStateException("No valid address book found in DB"));
+                .orElseGet(() -> {
+                    log.debug("No address book found in db");
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:124]: Log can be moved into migrate to produce a more accurate log for race condition and so you can shorten this to `.orElseGet(this::migrate)`. It should also be changed to info log level. In fact, all logs inside migrate can be info (besides the skipping migration log) since it should only occur once and is very important we understand what is going on.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 11]
```diff
@@ -111,13 +137,39 @@ public boolean isAddressBook(EntityId entityId) {
...
+     * converted to AddressBook objects and used to populate address_book and address_book_entry tables
+     *
+     * @return Latest AddressBook from historical files
+     */
+    private synchronized AddressBook migrate() {
+        if (addressBookRepository.count() > 0) {
+            if (log.isTraceEnabled()) {
+                log.trace("Valid address books exist in db, skipping migration");
+            }
+            return null;
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:151]: We should never return null. It will NPE during the race condition scenario. Instead of `count()`, can just call `addressBookRepository.findLatestAddressBook(102)` again like in `getCurrent()` and return it if not null. This is effectively double checked locking, so the same check should be done twice.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+@Log4j2
+public abstract class MirrorBaseJavaMigration extends BaseJavaMigration {
+
+    public abstract void doMigrate() throws Exception;
+
+    @Override
+    public void migrate(Context context) throws Exception {
+        MigrationVersion current = getVersion();
+        if (skipMigrationVersion(current, context.getConfiguration())) {
+            log.trace("Migration {} will be skipped as it precedes baseline version {}",
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java:38]: This will log as `MirrorBaseJavaMigration` instead of the concrete class. Should remove `@Log4j2` and use `protected final Logger log = LogManager.getLogger(getClass())` instead. You can also remove the annotation from children classes.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+import lombok.extern.log4j.Log4j2;
+import org.flywaydb.core.api.MigrationVersion;
+import org.flywaydb.core.api.configuration.Configuration;
+import org.flywaydb.core.api.migration.BaseJavaMigration;
+import org.flywaydb.core.api.migration.Context;
+
+@Log4j2
+public abstract class MirrorBaseJavaMigration extends BaseJavaMigration {
+
+    public abstract void doMigrate() throws Exception;
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java:32]: protected

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 14]
```diff
@@ -269,4 +321,68 @@ private void updatePreviousAddressBook(FileData fileData) {
...
+        AtomicLong currentConsensusTimestamp = new AtomicLong(0);
+        AtomicReference<AddressBook> lastAddressBook = new AtomicReference<>();
+
+        // starting from consensusTimeStamp = 0 retrieve pages of fileData entries for historic address books
+        int pageSize = 1000;
+        List<FileData> fileDataList = fileDataRepository
+                .findAddressBooksAfter(currentConsensusTimestamp.get(), pageSize);
+        while (!CollectionUtils.isEmpty(fileDataList)) {
+            log.info("Retrieved {} file_data rows for address book processing", fileDataList.size());
+            fileDataList.forEach(fileData -> {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:371]: nit: If you just used a normal for loop you could avoid all the atomics. ForEach is useful to be concise, but due to the atomics it actually makes things more verbose in my opinion.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 15]
```diff
@@ -269,4 +321,68 @@ private void updatePreviousAddressBook(FileData fileData) {
...
+                if (fileData.getFileData() != null && fileData.getFileData().length > 0) {
+                    // convert and ingest address book fileData contents
+                    lastAddressBook.set(parse(fileData));
+                    fileDataEntries.incrementAndGet();
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:375]: Codecov says `Added lines #L374 - L375 were not covered by tests`. Is that true? These seem like pretty important lines if not.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    /**
+     * Determine whether a java migration should be skipped based on version and isIgnoreMissingMigrations setting
+     *
+     * @param current                The current java migration version
+     * @param migrationConfiguration flyway Configuration
+     * @return
+     */
+    private boolean skipMigrationVersion(MigrationVersion current, Configuration migrationConfiguration) {
+        MigrationVersion baselineVersion = migrationConfiguration.getBaselineVersion();
+        if (baselineVersion.isNewerThan(current.getVersion()) && migrationConfiguration.isIgnoreMissingMigrations()) {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java:56]: Is it possible to get coverage for this?

Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java:56]: Why do we care about ignoreMissingMigrations here? Missing migrations is for files that have been deleted since they've been ran. I don't think that applies here. If you're baselining you're moving up the min version without regard for the presence of the migrations before that.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 9 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:125]: There shouldn't be. 
`migrate()` keeps track of the latest addressBook encountered, so it should return the same value as a `findLatestAddressBook()`
`migrate()` transactionTemplate also checks for empty historical case and ensures at the least the initialAddressBook is passed.
Only way that would fail is with a TransactionException and we'd wanna see those.
Let me know if that covers your question or if I'm missing something

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java (hunk 12 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java:38]: Good point

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 9 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:125]: It can return null on race condition between two threads migrating at same time.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 11 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:151]: Was siding with going with `count()` as an optimization but the double check is the more correct option

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 15 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:375]: I had a specific migration test(). It got removed in some refactoring.
Will add back as it definitely covers this

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java (hunk 16 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java:56]: Can do

Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java:56]: Will remove.
Also adding target logic as a nice to have

Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java]: Done

Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java]: Done

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 14 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Changed to a normal for loop

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java (hunk 13 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java (hunk 12 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/MirrorBaseJavaMigration.java:39]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 11 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 10 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 9 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Gotcha, with recent updated to check for current and return on non-null I believe that case should now be handled

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/V0_0_1__MirrorBaseJavaMigrationTest.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+// flyway requires class name conforms with naming convention
+class V0_0_1__MirrorBaseJavaMigrationTest extends MirrorBaseJavaMigration {
+    @Override
+    public void doMigrate() {
+        // do nothing
+    }
+
+    @Test
+    void verifySkipMigrationVersionCurrentEmpty() {
+        assertThat(skipMigrationVersion(MigrationVersion.EMPTY, getConfiguration("1", "2"))).isTrue();
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/V0_0_1__MirrorBaseJavaMigrationTest.java:38]: This should test migrate's invocation of skipMigrationVersion both to ensure it's correctly invoked and provide coverage to that method.

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/V0_0_1__MirrorBaseJavaMigrationTest.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+ */
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.flywaydb.core.api.MigrationVersion;
+import org.flywaydb.core.api.configuration.ClassicConfiguration;
+import org.junit.jupiter.api.Test;
+
+// flyway requires class name conforms with naming convention
+class V0_0_1__MirrorBaseJavaMigrationTest extends MirrorBaseJavaMigration {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/V0_0_1__MirrorBaseJavaMigrationTest.java:30]: Should just be named `MirrorBaseJavaMigrationTest`. You should create a custom private nested class to do the actual testing against.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_11_6__Missing_Entities.java [hunk 19]
```diff
@@ -54,14 +53,14 @@
...
-            @Lazy JdbcOperations jdbcOperations) {
+                                     @Lazy JdbcOperations jdbcOperations) {
         this.mirrorProperties = mirrorProperties;
         this.entityRepository = entityRepository;
         this.jdbcOperations = jdbcOperations;
     }
 
     @Override
-    public void migrate(Context context) throws Exception {
+    public void doMigrate() throws SchemaMigrationException {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_11_6__Missing_Entities.java:63]: You don't need to declare runtime exceptions (as Sonar also attests). Can just do `throws FileNotFoundException, IOException` as the wrapped exception doesn't provide any value since it's bubbled up to a framework we don't control.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java [hunk 20]
```diff
@@ -111,13 +132,50 @@ public boolean isAddressBook(EntityId entityId) {
...
+                .orElse(null);
+
+        if (currentAddressBook != null) {
+            if (log.isTraceEnabled()) {
+                log.trace("Valid address books exist in db, skipping migration");
+            }
+            return currentAddressBook;
+        }
+
+        log.debug("No address book found in db, proceeding with migration");
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java:155]: These two debugs should be info since this is ran once per lifetime.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java (hunk 20 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java]: Made info

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_11_6__Missing_Entities.java (hunk 19 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/V1_11_6__Missing_Entities.java]: Updated it to `IOException` as that's what's thrown and `FileNotFoundException` inherits from `IOException`.
This also requires the abstract class be updated with the throws

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/V0_0_1__MirrorBaseJavaMigrationTest.java (hunk 18 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/V0_0_1__MirrorBaseJavaMigrationTest.java]: Done

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/V0_0_1__MirrorBaseJavaMigrationTest.java (hunk 17 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/V0_0_1__MirrorBaseJavaMigrationTest.java]: Done

**Review by steven-sheehy — APPROVED** [review]
LGTM
